pisces/models.py

```diff
@@ -25,42 +27,123 @@
+        examples_X = [pair[0] for pair in pairs_Xy]
+        examples_y = [pair[1] for pair in pairs_Xy]
+
+
+    Xs = np.concatenate(examples_X, axis=0)
+    ys = np.concatenate(examples_y, axis=0)
+
+    selector = ys >= 0
+    Xs = Xs[selector]
+    ys = ys[selector]
+
+    loss_list = []
+    old_stdout = sys.stdout
+    sys.stdout = mystdout = StringIO()
+    classifier.pipeline.fit(Xs, ys) # Fit the model
+    sys.stdout = old_stdout
+    loss_history = mystdout.getvalue()
+    # Get loss
+    try:
+        for line in loss_history.split('\n'):
+            if(len(line.split("loss: ")) == 1):
+                continue
+            loss_list.append(float(line.split("loss: ")[-1]))
+    except:
+        warnings.warn("Failed to fetch loss history. Returning empty list.")
+
+    return loss_list
 
 # %% ../nbs/02_models.ipynb 8
-class SGDLogisticRegression(SleepWakeClassifier):
-    """Uses Sk-Learn's `SGDCLassifier` to train a logistic regression model. The SGD aspect allows for online learning, or custom training regimes through the `partial_fit` method.
-     
+@patch_to(SleepWakeClassifier)
+def train(self, examples_X: List[np.ndarray]=[], 
+          examples_y: List[np.ndarray]=[], 
+          pairs_Xy: List[Tuple[np.ndarray, np.ndarray]]=[],
+          ):
+    """
+    Assumes data is already preprocessed using `get_needed_X_y` 
+    and ready to be passed to the model.
+
+    Returns the loss history of the model.
+    """
+    loss_list = train_pipeline(self, examples_X, examples_y, pairs_Xy)
+    return loss_list
+
+# %% ../nbs/02_models.ipynb 10
+class LinearModel(Enum):
+    """Defines the loss used in sklearn's SGDClassifier which defines the linear model used for classification."""
+    LOGISTIC_REGRESSION = 'log_loss'
+    PERCEPTRON = 'perceptron'
+    SVM = 'hinge'
+
+# %% ../nbs/02_models.ipynb 11
+class SGDLinearClassifier(SleepWakeClassifier):
+    """Uses Sk-Learn's `SGDCLassifier` to train a model. Possible models are logistic regression, perceptron, and SVM.
+    The SGD aspect allows for online learning, or custom training regimes through the `partial_fit` method.
     The model is trained with a balanced class weight, and uses L1 regularization. The input data is scaled with a `StandardScaler` before being passed to the model.
     """
     def __init__(self, 
                  data_processor: DataProcessor, 
-                 lr: float = 0.15, 
-                 epochs: int = 100,):
-        self.model = SGDClassifier(loss='log_loss',
-                                   learning_rate='adaptive',
-                                   penalty='l1',
-                                   eta0=lr,
-                                   class_weight='balanced',
-                                   max_iter=epochs,
-                                   warm_start=True,
-                                   verbose=1)
+                 linear_model: LinearModel=LinearModel.LOGISTIC_REGRESSION,
+                 **kwargs):
```

Comment:
since you refactored the ABC to have these lines, I think we can replace this init with
```python
super().__init__(
    model=SGDClassifier(loss=linear_model.value, **kwargs),
    data_processor=data_processor)
if not isinstance(data_processor.model_input, ModelInput1D):
    raise ValueError("Model input must be set to 1D on the data processor")
if not data_processor.output_type == ModelOutputType.SLEEP_WAKE:
    raise ValueError("Model output must be set to SleepWake on the data processor")